### PR TITLE
Return Interceptor listeners to the caller

### PR DESCRIPTION
**Pull request by hexploitable**
Addresses #13.

At first, I thought an array of listeners would be enough, at least it is for my use-case. But perhaps there is a wider desire to have some context around the listeners returned, so I decided to return a dictionary. Let me know if you prefer this done another way.
Not convinced I'm handling the vtable stuff in a sensible way either.

Otherwise, this should be a trivial change.

### CODE REVIEW

**Comment by hexploitable** [comment]
Should now be good for review/close :)

File: index.js [hunk 1]
```diff
@@ -10,6 +10,7 @@ function trace (spec) {
   const {module, vtable, functions} = spec;
   const {onError} = spec.callbacks;
 
+  let listeners = Array();
```
Comment (oleavr) [inline index.js:13]:
```suggestion
  const listeners = [];
```

- Always use `const` when a variable is assigned just once, `let` should only be used when re-assigning is needed.
- Prefer array literals over invoking the constructor.

File: index.js [hunk 2]
```diff
@@ -43,13 +44,20 @@ function trace (spec) {
...
-      intercept(entry, impl);
+      listeners.push(intercept(entry, impl));
 
       offset += pointerSize;
     }
   } else {
     throw new Error('Either a module or a vtable must be specified');
   }
+  return {
+    stop: function() {
```
Comment (oleavr) [inline index.js:55]:
```suggestion
    stop() {
```

File: index.js [hunk 3]
```diff
@@ -43,13 +44,20 @@ function trace (spec) {
...
+      listeners.forEach((listener) => {
```
Comment (oleavr) [inline index.js:56]:
```suggestion
      listeners.forEach(listener => {
```

File: index.js [hunk 4]
```diff
@@ -43,13 +44,20 @@ function trace (spec) {
...
+        listener.detach();
+      })
```
Comment (oleavr) [inline index.js:58]:
```suggestion
      });
```

File: index.js [hunk 5]
```diff
@@ -101,7 +109,7 @@ function makeInterceptor (spec) {
     const numInputActions = inputActions.length;
     const numOutputActions = outputActions.length;
 
-    Interceptor.attach(impl, {
+    return Interceptor.attach(impl, {
```
Comment (oleavr) [inline index.js:112]: Here we return an object with `detach()` instead of one with `stop()`. Might be worth adding an internal class named something like `Session` which we return in both cases, and which can hold the `listeners`.

File: index.js [hunk 6]
```diff
@@ -43,13 +44,20 @@ function trace (spec) {
...
 
-      intercept(entry, impl);
+      listeners.push(intercept(entry, impl));
 
       offset += pointerSize;
     }
   } else {
     throw new Error('Either a module or a vtable must be specified');
   }
+  return {
```
Comment (oleavr) [inline index.js:54]:
```suggestion

  return {
```

File: index.js (hunk 5 above)
Comment (hexploitable) [reply index.js:106]: after some discussion offline, we've settled on the answer

File: index.js [hunk 7]
```diff
@@ -422,3 +424,15 @@ class Event {
     }
   }
 }
+
+class Session {
+  constructor (listeners) {
+    this.listeners = listeners;
+  }
+
+  stop() {
```
Comment (oleavr) [inline index.js:434]:
```suggestion
  stop () {
```

(Seeing as we're following semistandard in this project.)

File: index.js [hunk 8]
```diff
@@ -422,3 +424,15 @@ class Event {
...
+    this.listeners.forEach(listener => {
+      listener.detach();
+    })
```
Comment (oleavr) [inline index.js:437]:
```suggestion
    });
```

**Comment by hexploitable** [comment]
👍

File: index.js [hunk 9]
```diff
@@ -422,3 +424,15 @@ class Event {
     }
   }
 }
+
+class Session {
+  constructor (listeners) {
+    this.listeners = listeners;
```
Comment (oleavr) [inline index.js:431]:
```suggestion
    this._listeners = listeners;
```

To make it clear that this isn't public API.
